src/plugins/discover/public/application/components/context_app/use_context_app_state.ts

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+import { useEffect, useMemo, useState } from 'react';
+
+import { cloneDeep } from 'lodash';
+import { CONTEXT_DEFAULT_SIZE_SETTING } from '../../../../common';
+import { IndexPattern } from '../../../../../data/public';
+import { DiscoverServices } from '../../../build_services';
+import { AppState, getState } from '../../angular/context_state';
+
+export function useContextAppState({
+  indexPattern,
+  services,
+}: {
+  indexPattern: IndexPattern;
+  services: DiscoverServices;
+}) {
+  const { uiSettings: config, history, core, filterManager } = services;
+
+  const stateContainer = useMemo(() => {
+    return getState({
+      defaultSize: parseInt(config.get(CONTEXT_DEFAULT_SIZE_SETTING), 10),
+      timeFieldName: indexPattern.timeFieldName as string,
+      storeInSessionStorage: config.get('state:storeInSessionStorage'),
+      history: history(),
+      toasts: core.notifications.toasts,
+      uiSettings: config,
+    });
+  }, [config, history, indexPattern, core.notifications.toasts]);
+
+  const [appState, setState] = useState<AppState>(stateContainer.appState.getState());
+
+  /**
+   * Sync with app state container
+   */
+  useEffect(() => {
+    stateContainer.startSync();
+
+    return () => stateContainer.stopSync();
+  }, [stateContainer]);
+
+  useEffect(() => {
+    const unsubscribeAppState = stateContainer.appState.subscribe((newState) => {
+      stateContainer.flushToUrl(true);
```

Comment:
This functional **[test](https://github.com/Dmitriynj/kibana/blob/master/test/functional/apps/context/_context_navigation.js#L51)** needs flushing url. The line `await browser.goBack();` there will not gonna do redirect to discover page if we do not flushing the url

Reply:
Had a closer look at it, works, but with a few changes calling this function could be omitted (it was implemented for testing). So in context we do not want to  push to history like in Discover Main, we want to replace in history. If we would change the number of successor, predecessors 3 times, we would need 4 back button clicks to come back to Discover Main.

This can be solved in `context_state.ts` by 2 changes like

Line 192
```
stateStorage.set(APP_STATE_URL_KEY, mergedState, { replace: true });
```

Line 207
```
stateStorage.set(GLOBAL_STATE_URL_KEY, { filters: globalFilters }, { replace: true });
```

Line 216
```
const nextState = { ...appStateContainer.getState(), ...{ filters: appFilters } };
stateStorage.set(APP_STATE_URL_KEY, nextState, { replace: true });
```
Then 
```
stateContainer.flushToUrl(true);
```
will no longer be necessary 🥳

Reply:
Checked again in the old version, and there filters are pushed to history, not replaced. So the only change necessary should be in Line 192